modules/common/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/common/transformation/parquet/NonAtomicFieldsProvider.scala

```diff
@@ -6,40 +6,142 @@ import cats.effect.Clock
 import com.snowplowanalytics.snowplow.badrows.FailureDetails.LoaderIgluError
 import com.snowplowanalytics.snowplow.rdbloader.common.LoaderMessage.TypesInfo.WideRow
-import com.snowplowanalytics.snowplow.rdbloader.common.LoaderMessage.TypesInfo.WideRow.latestByModel
 import com.snowplowanalytics.snowplow.rdbloader.common.transformation.parquet.fields.{NonAtomicFields, TypedField}
+import com.snowplowanalytics.snowplow.rdbloader.common.SchemaProvider._
 import com.snowplowanalytics.snowplow.rdbloader.common.{LoaderMessage, SchemaProvider}
 
+import scala.math.Ordered.orderingToOrdered
+import scala.math.abs
+
 object NonAtomicFieldsProvider {
 
+  /**
+   * Builds a `NonAtomicFields`.
+   * @param resolver
+   *   Iglu resolver, that would be used to fetch the schema list and their content for the `types`.
+   * @param types
+   *   list of schemas types (SchemaKey and Struct/Unstruct flag) for the single event
+   * @tparam F
+   *   \- IO
+   * @return
+   *   List of unique TypedFields. See `extractEndSchemas` docstring for explanation of TypedFields.
+   */
   def build[F[_]: Clock: Monad: RegistryLookup](
     resolver: Resolver[F],
     types: List[WideRow.Type]
   ): EitherT[F, LoaderIgluError, NonAtomicFields] =
-    latestByModel(types).sorted
-      .traverse(createTypedField(resolver))
+    types.sorted
+      // collapse subversions of the same schema. This will avoid listing the earlier schema versions multiple times.
+      .foldRight(List.empty[WideRow.Type])((it, acc) =>
+        acc match {
+          case Nil => List(it)
+          case head :: _ =>
+            if (
+              head.snowplowEntity == it.snowplowEntity &
+                head.schemaKey.vendor == it.schemaKey.vendor &
+                head.schemaKey.name == it.schemaKey.name &
+                head.schemaKey.format == it.schemaKey.format &
+                head.schemaKey.version.model == it.schemaKey.version.model
+            )
+              acc
+            else
+              it :: acc
+        }
+      )
+      .flatTraverse(fieldFromType(resolver))
       .map(NonAtomicFields)
 
-  private def createTypedField[F[_]: Clock: Monad: RegistryLookup](
-    resolver: Resolver[F]
-  )(
-    `type`: WideRow.Type
-  ): EitherT[F, LoaderIgluError, TypedField] =
-    fieldFromType[F](resolver, `type`)
-      .map(field => TypedField(field, `type`))
+  /**
+   * Extract TypedFields for the `type`. It could produce multiple TypedFields for the same
+   * `type`.schema. If output is longer than one item, it means that user created a broken
+   * migration. Broken migration constitutes a illegal type casting and defined in schema-dll
+   * package accessed by `isSchemaMigrationBreaking`.
+   *
+   * For example, changing field type from integer to string between 1-0-0 and 1-0-1 versions. That
+   * example, would produce two Fields (in pseudocode):
+   *   - TypedField( `type`=( schemaKey = iglu://my_iglu_schema/1-0-0 ), lowerBound = None, Field=(
```

Comment:
We have one 'integration' like spec (which is great!) in batch/streaming transformer for this feature. What do you think about adding more unit tests for `build` function? I think it should relatively easy in e.g. `ParquetFieldsProviderSpec`? 

I suspect there are a lot of different cases to consider and if we have tests covering them it would help to visualize/understand this feature better! 

Idea: could we use 'in memory' version of iglu registry in such tests to not clutter  resources dir...?

Reply:
> What do you think about adding more unit tests for build function? I think it should relatively easy in e.g. ParquetFieldsProviderSpec?

I like this idea. Just need to figure out if we are going with @istreeter suggestion to use `modifiedTime` ordering. What is your opinion on it?

>could we use 'in memory' version of iglu registry in such tests to not clutter resources dir...?

The main issue is that would require adding a list schema feature to the in-memory repo. I've already done it for the embedded.
Secondly, it wouldn't match the test style for the rest of the tests in the same spec. I think consistency should also be considered.

Reply:
How hard would that be to do it also for in-memory? If it's too much fuss, then forget about it. But if you're currently working on it and it's pretty easy maybe it's worth doing it? 

I don't mind changing the style in the spec if the new style makes those tests more readable and makes adding new test cases easier ;) So if working with embedded repo doesn't cause any problems, we can leave it as is!